### PR TITLE
cGAN tutorial for MXNet R package

### PR DESCRIPTION
**Pull request by jeremiedb**
Following @thirdwing proposal, here is a blog post on training a cGAN network with R package. 
Not sure if the image links was done properl, example: 
``
Could you please validate?

### CODE REVIEW

**Comment by thirdwing** [comment]
I suggest to host the figures at https://github.com/dmlc/web-data, like ``

**Comment by jeremiedb** [comment]
Made the change and made pulled request to web-data.

**Comment by thirdwing** [comment]
@hetong007 @tqchen 

Please take a look and leave some comments.

File: _posts/2017-06-01-Generative-Adversial-Network-in-R.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+The following packages are needed to run the tutorial: 
+
+```r
+require("imager")
+require("dplyr")
+require("readr")
+require("mxnet")
+```
+
+The full demo is comprised of the two following scripts available on [GitHub](https://github.com/jeremiedb/gan_example):
```
Comment (thirdwing) [inline _posts/2017-06-01-Generative-Adversial-Network-in-R.md]: @jeremiedb Do you mind sending a PR to put the code into https://github.com/dmlc/mxnet/tree/master/example

File: _posts/2017-06-01-Generative-Adversial-Network-in-R.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+![](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/blog_R_cgan/Discriminator.png)
+
+### Training logic
+
+The training process of the discriminator is most obvious: the loss is simple a binary TRUE/FALSE response and that loss is propagated back into the CNN network. It can therefore be understood as a simple binary classification problem. 
+
+```r
+### Train loop on fake
+mx.exec.update.arg.arrays(exec_D, arg.arrays = list(data=D_data_fake, digit=D_digit_fake, label=mx.nd.array(rep(0, batch_size))), match.name=TRUE)
```
Comment (thirdwing) [inline _posts/2017-06-01-Generative-Adversial-Network-in-R.md]: The code seems too long to be in one line. Can you change line in the middle?

File: _posts/2017-06-01-Generative-Adversial-Network-in-R.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+mx.exec.update.arg.arrays(exec_D, update_args_D, skip.null=TRUE)
+```
+
+The generator loss comes from the backpropagation of the the discriminator loss into its generated output. By faking the generator labels to be real samples into the discriminator, the discriminator back-propagated loss provides the generator with the information on how to best adapt its parameters to trick the discriminator into believing the fake samples are real. 
+
+This requires to backpropagate the gradients up to the input data of the discriminator (whereas this input gradient is typically ignored in vanilla feedforward network).  
+
+```r
+### Update Generator weights - use a seperate executor for writing data gradients
+exec_D_back<- mxnet:::mx.symbol.bind(symbol = D_sym, arg.arrays = exec_D$arg.arrays, aux.arrays = exec_D$aux.arrays, grad.reqs = rep("write", length(exec_D$arg.arrays)), ctx = devices)
```
Comment (thirdwing) [inline _posts/2017-06-01-Generative-Adversial-Network-in-R.md]: The same.

File: _posts/2017-06-01-Generative-Adversial-Network-in-R.md (hunk 1 above)
Comment (hetong007) [reply _posts/2017-06-01-Generative-Adversial-Network-in-R.md]: Maybe under https://github.com/dmlc/mxnet/tree/master/example/gan

**Comment by hetong007** [comment]
LGTM!

File: _posts/2017-06-01-Generative-Adversial-Network-in-R.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+- `CGAN_mnist_setup.R`: prepare data and define the model structure
+- `CGAN_train.R`: execute the training
+
+### Data preperation
+
+The MNIST dataset is available on [Kaggle](https://www.kaggle.com/c/digit-recognizer/data). Once `train.csv` is downloaded into the `data/` folder, we can import into R.  
+
+```r
+train <- read_csv('data/train.csv')
+train<- data.matrix(train)
```
Comment (hetong007) [inline _posts/2017-06-01-Generative-Adversial-Network-in-R.md]: Could you please have space before `<-` as well? I see several similar cases below.

**Review by hetong007 — APPROVED** [review]

**Comment by thirdwing** [comment]
Let's merge the code into mxnet repo first, then merge the blog.

**Comment by thirdwing** [comment]
@tqchen  Please take a look and give some comments.

**Comment by tqchen** [comment]
LGTM!

**Comment by thirdwing** [comment]
@tqchen I am afraid I don't have the permission to merge this.

**Comment by BruceZhaoR** [comment]
@thirdwing   I found a typo, `Adversial` should be ` Adversarial`.   I have added the post link to the [Rweekly  draft](https://github.com/rweekly/rweekly.org/blob/gh-pages/draft.md#r-in-the-real-world), which will be released soon.  If you would like to change the link address , please let me know , or send a PR directly, thanks !

**Comment by thirdwing** [comment]
@BruceZhaoR Sorry for the typo, please send a PR.

**Pull request merged**